File: core-api/app/core/invite_token.py
import hmac
import hashlib
import json
import base64
from datetime import datetime, timezone
from typing import Optional
from app.core.config import get_settings


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def _b64url_decode(s: str) -> bytes:
    pad = 4 - len(s) % 4
    if pad != 4:
        s += "=" * pad
    return base64.urlsafe_b64decode(s)


def _sign(data: str) -> str:
    secret = get_settings().invite_token_secret.encode()
    return _b64url_encode(
        hmac.new(secret, data.encode(), hashlib.sha256).digest()
    )


def create_invite_token(
    email: str,
    tenant_id: str,
    role: str,
    invited_by: str,
    expires_at: str,
) -> str:
    # Field order must match TS: {email, tenantId, role, invitedBy, expiresAt}
    # separators=(",", ":") removes whitespace so base64url matches TS output
    payload = {"email": email, "tenantId": tenant_id, "role": role,
               "invitedBy": invited_by, "expiresAt": expires_at}
    data = _b64url_encode(json.dumps(payload, separators=(",", ":")).encode())
    sig = _sign(data)
    return f"{data}.{sig}"


def verify_invite_token(token: str) -> Optional[dict]:
    try:
        dot = token.rfind(".")
        if dot == -1:
            return None
        data, sig = token[:dot], token[dot + 1:]
        expected = _sign(data)
        if not hmac.compare_digest(sig, expected):
            return None
        payload = json.loads(_b64url_decode(data))
        # Python 3.9 fromisoformat() doesn't handle trailing 'Z' — replace it with +00:00
        expires_str = payload["expiresAt"].replace("Z", "+00:00")
        if datetime.fromisoformat(expires_str) < datetime.now(timezone.utc):
            return None
        return payload
    except Exception:
        return None

File: apps/revenue-retention-os/src/lib/invite-token.ts
// HMAC-SHA256 기반 초대 토큰 서명/검증
// 환경변수 INVITE_TOKEN_SECRET 필요 (최소 32자 랜덤 문자열)

import { createHmac, timingSafeEqual } from 'crypto'
import type { UserRole } from '@/types/domain'

export interface InviteTokenPayload {
  email: string
  tenantId: string
  role: UserRole
  invitedBy: string
  expiresAt: string
}

function getSecret(): string {
  const secret = process.env.INVITE_TOKEN_SECRET
  if (!secret && process.env.NODE_ENV === 'production') {
    throw new Error('INVITE_TOKEN_SECRET 환경변수가 설정되지 않았습니다')
  }
  return secret || 'dev-secret-change-in-production'
}

function sign(payload: string): string {
  return createHmac('sha256', getSecret())
    .update(payload)
    .digest('base64url')
}

export function createInviteToken(payload: InviteTokenPayload): string {
  const data = Buffer.from(JSON.stringify(payload)).toString('base64url')
  const sig  = sign(data)
  return `${data}.${sig}`
}

export function verifyInviteToken(token: string): InviteTokenPayload | null {
  try {
    const dot = token.lastIndexOf('.')
    if (dot === -1) return null

    const data     = token.slice(0, dot)
    const sig      = token.slice(dot + 1)
    const expected = sign(data)

    // 타이밍 공격 방지
    if (!timingSafeEqual(Buffer.from(sig), Buffer.from(expected))) return null

    const payload = JSON.parse(
      Buffer.from(data, 'base64url').toString()
    ) as InviteTokenPayload

    if (new Date(payload.expiresAt) < new Date()) return null
    return payload
  } catch {
    return null
  }
}
